examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,465 @@
+    PetscCall(PetscFEDestroy(&fe_coord));
+  }
+
+  // -- Set tensor permutation
+  {
+    DM dm_coord;
+
+    PetscCall(DMGetCoordinateDM(dm_mesh, &dm_coord));
+    PetscCall(DMPlexSetClosurePermutationTensor(dm_mesh, PETSC_DETERMINE, NULL));
+    // PetscCall(DMPlexSetClosurePermutationTensor(dm_coord, PETSC_DETERMINE, NULL));
+  }
+
+  // -- Final background mesh
+  PetscCall(PetscObjectSetName((PetscObject)dm_mesh, "Background Mesh"));
+  PetscCall(DMViewFromOptions(dm_mesh, NULL, "-dm_mesh_view"));
+
+  // -- libCEED objects from background mesh
+  {
+    BPData bp_data = {.q_mode = CEED_GAUSS};
+
+    PetscCall(CreateBasisFromPlex(ceed, dm_mesh, NULL, 0, 0, 0, bp_data, &basis_mesh_u));
+    PetscCall(CreateRestrictionFromPlex(ceed, dm_mesh, 0, NULL, 0, &restriction_mesh_u));
+  }
+
+  // Create particle swarm
+  PetscCall(DMCreate(comm, &dm_swarm));
+  PetscCall(DMSetType(dm_swarm, DMSWARM));
+  PetscCall(DMSetDimension(dm_swarm, dim));
+  PetscCall(DMSwarmSetType(dm_swarm, DMSWARM_PIC));
+  PetscCall(DMSwarmSetCellDM(dm_swarm, dm_mesh));
+
+  // -- Swarm field
+  PetscCall(DMSwarmRegisterPetscDatatypeField(dm_swarm, "u", num_comp, PETSC_SCALAR));
+  PetscCall(DMSwarmFinalizeFieldRegister(dm_swarm));
+  PetscCall(DMSwarmSetLocalSizes(dm_swarm, num_points, 0));
+  PetscCall(DMSetFromOptions(dm_swarm));
+
+  // -- Set swarm point locations
+  {
+    PetscScalar *point_coords;
+
+    PetscCall(DMSwarmGetField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&point_coords));
+    for (PetscInt p = 0; p < num_points; p++) {
+      point_coords[p * dim + 0] = -PetscCosReal((PetscReal)(p + 1) / (PetscReal)(num_points + 1) * PETSC_PI);
+      if (dim > 1) point_coords[p * dim + 1] = -PetscSinReal((PetscReal)(p + 1) / (PetscReal)(num_points + 1) * PETSC_PI);
+      if (dim > 2) point_coords[p * dim + 2] = PetscSinReal((PetscReal)(p + 1) / (PetscReal)(num_points + 1) * PETSC_PI);
+    }
+    PetscCall(DMSwarmRestoreField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&point_coords));
+  }
+  PetscCall(DMSwarmMigrate(dm_swarm, PETSC_TRUE));
+
+  // -- Final particle swarm
+  PetscCall(PetscObjectSetName((PetscObject)dm_swarm, "Particle Swarm"));
+  PetscCall(DMViewFromOptions(dm_swarm, NULL, "-dm_swarm_view"));
+
+  // Set field values on background mesh
+  PetscCall(DMCreateGlobalVector(dm_mesh, &U_mesh));
+  {
+    Vec                X_loc, U_loc;
+    PetscInt           num_mesh_points_loc;
+    PetscScalar       *u_array;
+    const PetscScalar *x_array;
+
+    PetscCall(DMGetCoordinatesLocal(dm_mesh, &X_loc));
+    PetscCall(DMGetLocalVector(dm_mesh, &U_loc));
+    PetscCall(VecGetLocalSize(U_loc, &num_mesh_points_loc));
+    PetscCall(VecGetArray(U_loc, &u_array));
+    PetscCall(VecGetArrayRead(X_loc, &x_array));
+    for (PetscInt i = 0; i < num_mesh_points_loc; i++) {
+      PetscScalar x[dim];
+
+      for (PetscInt d = 0; d < dim; d++) x[d] = x_array[d * num_mesh_points_loc + i];
```

Comment:
Our coordinate ordering is wrong here, it should be:
```suggestion
      for (PetscInt d = 0; d < dim; d++) x[d] = x_array[i * dim + d];
```